src/dataflow/src/source/kafka.rs

```diff
@@ -437,48 +444,64 @@ impl DataPlaneInfo {
-            let partition_queue = self
-                .consumer
-                .split_partition_queue(&self.topic_name, partition_id);
-            if let Some(partition_queue) = partition_queue {
-                self.partition_consumers
-                    .push_front(PartitionConsumer::new(partition_id, partition_queue));
-                assert_eq!(
-                    self.consumer
-                        .assignment()
-                        .unwrap()
-                        .elements_for_topic(&self.topic_name)
-                        .len(),
-                    self.partition_consumers.len()
-                );
-                self.partition_metrics.insert(
-                    partition_id,
-                    PartitionMetrics::new(
-                        &self.topic_name,
-                        &self.source_id,
-                        &partition_id.to_string(),
-                    ),
-                );
-                self.consumer.poll(Duration::from_secs(0));
-                true
-            } else {
-                false
-            }
+        // Collect old partition assignments
+        let tpl = self.consumer.assignment().unwrap();
+        // Create list from assignments
+        let mut partition_list = TopicPartitionList::new();
+        for partition in tpl.elements_for_topic(&self.topic_name) {
+            partition_list.add_partition_offset(
+                partition.topic(),
+                partition.partition(),
+                partition.offset(),
+            );
+        }
+        // Add new partition
+        partition_list.add_partition_offset(&self.topic_name, partition_id, Offset::Beginning);
+        self.consumer.assign(&partition_list).unwrap();
+
+        // Trick librdkafka into updating its metadata for the topic so that we
+        // start seeing data for the partition immediately. Otherwise we might
+        // need to wait the full `topic.metadata.refresh.interval.ms` interval.
+        //
+        // We don't actually care about the results of the metadata refresh, and
+        // would prefer not to wait for it to complete, so we execute this
+        // request with a timeout of 0s and ignore the result. This relies on an
+        // implementation detail, which is that librdkafka does not proactively
+        // cancel metadata fetch operations when they reach their timeout.
+        // Unfortunately there is no asynchronous metadata fetch API.
+        //
+        // It is not a problem if the metadata request fails, because the
+        // background metadata refresh will retry indefinitely. As long as a
+        // background request succeeds eventually, we'll start receiving data
+        // for the new partitions.
+        //
+        // TODO(benesch): remove this if upstream makes this metadata refresh
+        // happen automatically [0].
+        //
+        // [0]: https://github.com/edenhill/librdkafka/issues/2917
+        let _ = self
```

Comment:
Like I said in my comments, I think this risks (from what I've seen) causing us to wait until the background refresh more often than not for high partitions count. That might be a worthwhile trade-off to avoid the time-out, but wouldn't a better approach be to set the timeout to 0 so that we never block, but reschedule the operator  a second later to try again? 1 second should be enough to not hammer Kafka, but will also minimise startup delay?